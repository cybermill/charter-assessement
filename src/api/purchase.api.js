import { purchaseRecords } from "../mock.data";

export const fetchPurchaseRecords = () =>
  new Promise((res, rej) => {
    setTimeout(() => {
      res(purchaseRecords);
    }, 2000);
  });

// console.log("fetch", fetchPurchaseRecords);
