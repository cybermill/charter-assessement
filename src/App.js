/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState } from "react";
import { fetchPurchaseRecords } from "./api/purchase.api";
import PurchaseTable from "./components/PurchaseTable/PurchaseTable";
import { useLoading } from "./hooks/useLoading";

//how to group by months?
//how to group by customerID?
//username + customerID must be the same
//reward system: 1pt = $50; 2pt = $100
//track 3 months
//reward points/customer per purchase
//total reward points/customer per month

export const purchaseRecordsGroupByMonth = (purchaseRecords) => {
  const months = new Array(12).fill(0).map((_) => []);
  // console.log("months", months);
  purchaseRecords.forEach((record) => {
    const month = new Date(record.purchaseDate).getMonth();
    months[month].push(record);
  });
  return months;
};

export const purchaseRecordsGroupByCustomer = (purchaseRecords) => {
  const customers = {};
  purchaseRecords.forEach((record) => {
    const customerId = record.customerID;
    // console.log("record", record);
    if (!customers[customerId]) {
      customers[customerId] = [record];
    } else {
      customers[customerId].push(record);
    }
  });
  return customers;
};

export const purchaseRecordsParser = (purchaseRecords) => {
  let monthsRecords = purchaseRecordsGroupByMonth(purchaseRecords);
  // console.log("monthsRecords", monthsRecords);
  monthsRecords = monthsRecords.map((records) => {
    return purchaseRecordsGroupByCustomer(records);
  });
  return monthsRecords;
};

export default function App() {
  const [isLoading, startLoading, endLoading] = useLoading();
  const [monthlyRecordsByCustomer, setMonthlyRecordsByCustomer] = useState(
    null
  );
  // console.log("monthlyRecordsByCustomer", monthlyRecordsByCustomer);

  useEffect(() => {
    startLoading();
    fetchPurchaseRecords().then((data) => {
      setMonthlyRecordsByCustomer(purchaseRecordsParser(data));
      endLoading();
    });
  }, []);

  return (
    <div className="App">
      {isLoading ? (
        <h3>isLoading...</h3>
      ) : (
        monthlyRecordsByCustomer?.map((monthlyRecords, MonthNum) => {
          // console.log("monthlyRecords", monthlyRecords);
          const customersRecords = Object.keys(monthlyRecords);
          // console.log("customersRecords", customersRecords);
          if (customersRecords.length > 0) {
            return customersRecords.map((customerID) => {
              return (
                <section key={monthlyRecords[customerID][0].id}>
                  <PurchaseTable
                    purchaseRecords={monthlyRecords[customerID]}
                    monthNum={MonthNum}
                    customerID={customerID}
                    customerName={monthlyRecords[customerID][0].customerName}
                  />
                </section>
              );
            });
          }
          return null;
        })
      )}
    </div>
  );
}

//how to render UI with the current data structure?
//current data structure: [{}..., object..., {}...]
//month = { key: [ob,ob,...]}
