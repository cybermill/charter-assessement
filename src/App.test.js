import React from 'react';
import '@testing-library/jest-dom';
import { render, screen} from '@testing-library/react';
import App from './App';
import * as api from './api/purchase.api'
import { act } from 'react-dom/test-utils';
import { purchaseRecordsGroupByMonth, purchaseRecordsParser, purchaseRecordsGroupByCustomer } from "./App"

const mockData = [
    {
        id: 1,
        customerID: 123,
        customerName: "John",
        purchasePrice: 100,
        purchaseDate: new Date("May 17, 2022 03:24:00").toLocaleDateString()
    },
    {
        id: 2,
        customerID: 234,
        customerName: "Ashley",
        purchasePrice: 200,
        purchaseDate: new Date("June 17, 2022 03:24:00").toLocaleDateString()
    },
    {
        id: 3,
        customerID: 345,
        customerName: "Christina",
        purchasePrice: 300,
        purchaseDate: new Date("July 17, 2022 03:24:00").toLocaleDateString()
    }
]

describe('App', () => {
    beforeAll(() => {
        // we're using fake timers because we don't want to
        // wait a full second for this test to run.
        jest.useFakeTimers()
    })
    afterAll(() => {
        jest.useRealTimers()
    })
    it('should match snapshot', () => {
        const { asFragment } = render(<App />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should fetch purchase records when component did mount', async () => {
        jest.spyOn(api, 'fetchPurchaseRecords').mockResolvedValue([])
        render(<App />);
        await act(async () => jest.advanceTimersByTime(1000))
        expect(api.fetchPurchaseRecords).toBeCalledTimes(1);
    })
    it('should show loading when fetching the data', async () => {
        jest.spyOn(api, 'fetchPurchaseRecords').mockResolvedValue([])
        render(<App />);
        const paragraphElement = screen.getByText(/isLoading.../i);
        expect(paragraphElement).toBeInTheDocument();
        expect(paragraphElement.tagName).toBe('H3')
        await act(async () => jest.advanceTimersByTime(1000))
        expect(paragraphElement).not.toBeInTheDocument();
    });
    
    it('should return an array of arrays', () => {
        const output = [[],[],[],[],[mockData[0]],[mockData[1]],[mockData[2]],[],[],[],[],[]]
        expect(purchaseRecordsGroupByMonth(mockData)).toEqual(output);
        
    })

    it('should return an array of objects', () => {
        const output = [{},{},{},{},{ 123 : [ mockData[0] ] },{ 234: [ mockData[1] ] },{ 345: [ mockData[2] ]},{},{},{},{},{}]
        expect(purchaseRecordsParser(mockData)).toEqual(output);
    })

    it('should return an object with customerID as keys', () => {
        const output = { 
            123 : [
                    {   id: 1,
                        customerID: 123,
                        customerName: "John",
                        purchasePrice: 100,
                        purchaseDate: new Date("May 17, 2022 03:24:00").toLocaleDateString()
                    }
            ],
            234 : [
                    {   id: 2,
                        customerID: 234,
                        customerName: "Ashley",
                        purchasePrice: 200,
                        purchaseDate: new Date("June 17, 2022 03:24:00").toLocaleDateString()
                    }
            ],
            345 : [
                    {
                        id: 3,
                        customerID: 345,
                        customerName: "Christina",
                        purchasePrice: 300,
                        purchaseDate: new Date("July 17, 2022 03:24:00").toLocaleDateString()
                    }
            ]
        }
        expect(purchaseRecordsGroupByCustomer(mockData)).toEqual(output)
    })
});