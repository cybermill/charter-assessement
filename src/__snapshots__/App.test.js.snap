// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h3>
      isLoading...
    </h3>
  </div>
</DocumentFragment>
`;
